src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsTag.cs

```diff
@@ -0,0 +1,152 @@
+        public readonly InlineDiagnosticsLocations Location;
+        private readonly DiagnosticData _diagnostic;
+        private readonly INavigateToLinkService _navigateToLinkService;
+        private readonly IEditorFormatMap _editorFormatMap;
+
+        public InlineDiagnosticsTag(string errorType, DiagnosticData diagnostic, IEditorFormatMap editorFormatMap, InlineDiagnosticsLocations location, INavigateToLinkService navigateToLinkService)
+            : base(editorFormatMap)
+        {
+            ErrorType = errorType;
+            _diagnostic = diagnostic;
+            Location = location;
+            _navigateToLinkService = navigateToLinkService;
+            _editorFormatMap = editorFormatMap;
+        }
+
+        /// <summary>
+        /// Creates a GraphicsResult object which is the error block based on the geometry and formatting set for the item.
+        /// </summary>
+        public override GraphicsResult GetGraphics(IWpfTextView view, Geometry bounds, TextFormattingRunProperties format)
+        {
+            var block = new TextBlock
+            {
+                FontFamily = format.Typeface.FontFamily,
+                FontSize = 0.75 * format.FontRenderingEmSize,
+                FontStyle = FontStyles.Normal,
+                Foreground = format.ForegroundBrush,
+                Padding = new Thickness(left: 2, top: 0, right: 2, bottom: 0),
+            };
+
+            var id = new Run(_diagnostic.Id);
+            var link = new Hyperlink(id)
+            {
+                NavigateUri = new Uri(_diagnostic.HelpLink)
+            };
+
+            link.RequestNavigate += HandleRequestNavigate;
+            block.Inlines.Add(link);
+            block.Inlines.Add(": " + _diagnostic.Message);
+
+            var lineHeight = Math.Floor(format.Typeface.FontFamily.LineSpacing * block.FontSize);
+            var image = new CrispImage
+            {
+                Moniker = GetMoniker(),
+                MaxHeight = lineHeight,
+                Margin = new Thickness(1, 0, 5, 0)
+            };
+            var stackPanel = new StackPanel
+            {
+                Height = lineHeight,
+                Orientation = Orientation.Horizontal
+            };
+
+            stackPanel.Children.Add(image);
+            stackPanel.Children.Add(block);
+
+            var border = new Border
+            {
+                BorderBrush = format.BackgroundBrush,
+                BorderThickness = new Thickness(1),
+                Background = Brushes.Transparent,
+                Child = stackPanel,
+                CornerRadius = new CornerRadius(2),
+                // Highlighting lines are 2px buffer. So shift us up by one from the bottom so we feel centered between them.
+                Margin = new Thickness(10, top: 0, 0, bottom: 1),
+                Padding = new Thickness(1)
+            };
+
+            var editorBackground = (Color)_editorFormatMap.GetProperties("TextView Background")["BackgroundColor"];
```

Comment:
is `TextView Background` a constant that is sure to be there?  i'm slightly worried about localization doing something wonky with that.

Reply:
I believe it should always be there? When would it not?

Reply:
> When would it not?

No clue :D   I'm just always scared blindly indexing into objects we don't own.  Are things like 'TextView Background' also known to be strings that are safe to use in any local?  Do we have code like this anywhere else in roslyn?  Did someone on editor/platform indicate this is the right way to do this?

Reply:
I can ask the editor team if it is okay

Reply:
These should never be localized, they aren't user facing strings but indexes into resources. Doesn't hurt to ask the editor, but I'm 99.99% certain this won't change. It's basically indexing into a WPF defined ResourceDictionary

Reply:
ok.   seems workable for now :)